### PR TITLE
OCPBUGS-19700: Modified Route-specific annotations table and added a section for ingress controller annotation

### PR DESCRIPTION
**Pull request by subhtk**
Version(s): 4.12+

Issue: [OCPBUG 19700](https://issues.redhat.com/browse/OCPBUGS-19700)

Link to docs preview: [Doc preview](https://68972--ocpdocs-pr.netlify.app/openshift-enterprise/latest/networking/routes/route-configuration#nw-route-specific-annotations_route-configuration)

QE review:
- [x] QE has approved this change.

Additional information:
[Table](https://docs.google.com/document/d/1pDh3ubPgRTaxlwj-aMLmVlntCBMHXYPnV_2FE-kicBI/edit)

### CODE REVIEW

**Comment by palonsoro** [comment]
I am afraid it is wrong:
- Columns have to be "Ingress controller field" and "Ingress controller annotation"
- The values are not just copy-pasting the environment variables. They must contain the right ingress controller fields (where applicable) and the right ingress controller annotations (where applicable)
- The table about router timeouts still references the environment variables, which is wrong.

Basically this is not a simple rename, what is needed here is a full rework of the whole section.

**Comment by palonsoro** [comment]
lgtm

**Comment by subhtk** [comment]
@ShudiLi can you ptal?

File: modules/nw-route-specific-annotations.adoc [hunk 1]
```diff
@@ -16,55 +16,122 @@ To create a whitelist with multiple source IPs or subnets, use a space-delimited
...
+[cols="4*", options="header"]
 |===
-|Variable | Description | Environment variable used as default
-|`haproxy.router.openshift.io/balance`| Sets the load-balancing algorithm. Available options are `random`, `source`, `roundrobin`, and `leastconn`.  The default value is `source` for TLS passthrough routes. For all other routes, the default is `random`. |`ROUTER_TCP_BALANCE_SCHEME` for passthrough routes. Otherwise, use `ROUTER_LOAD_BALANCE_ALGORITHM`.
-|`haproxy.router.openshift.io/disable_cookies`| Disables the use of cookies to track related connections. If set to `'true'` or `'TRUE'`, the balance algorithm is used to choose which back-end serves connections for each incoming HTTP request. |
-|`router.openshift.io/cookie_name`| Specifies an optional cookie to use for
+|Variable 
+|Description 
+|Ingress controller field
+|Ingress controller annotation
```
Comment (Miciah) [inline modules/nw-route-specific-annotations.adoc:24]: As I suggested for "Ingress controller field", can you delete this column and incorporate the information in this column (for the one row that uses it) into the "Description" column?  

I cannot even see this column without scrolling, and so I didn't even know the table had this column until I looked at the AsciiDoc.  I am currently viewing the preview on a very large display, so I expect that most people who look at this table will not see the rightmost column and will not realize it exists.

File: modules/nw-route-specific-annotations.adoc [hunk 2]
```diff
@@ -16,55 +16,122 @@ To create a whitelist with multiple source IPs or subnets, use a space-delimited
...
+|
+
+|`haproxy.router.openshift.io/rate-limit-connections.rate-tcp`
+|Limits the rate at which a client with the same source IP address can make TCP connections. It accepts a numeric value.  +
+Note: Using this annotation provides basic protection against distributed denial-of-service (DDoS) attacks.
+|
+|
+
+|`haproxy.router.openshift.io/timeout` 
+|Sets a server-side timeout for the route. (TimeUnits)
```
Comment (Miciah) [inline modules/nw-route-specific-annotations.adoc:74]: Just writing "TimeUnits" here is confusing.  I see that "TimeUnits" is defined below, under the heading "Router timeout variables".  Would it be possible to add a more explicit reference, using a footnote or a link to the "Router timeout variables" heading?

File: modules/nw-route-specific-annotations.adoc [hunk 3]
```diff
@@ -16,55 +16,122 @@ To create a whitelist with multiple source IPs or subnets, use a space-delimited
...
+|
+
+|`haproxy.router.openshift.io/timeout` 
+|Sets a server-side timeout for the route. (TimeUnits) 
+|`spec.tuningOptions.serverTimeout` 
+|
+
+|`haproxy.router.openshift.io/timeout-tunnel` 
+|This timeout applies to a tunnel connection, for example, WebSocket over cleartext, edge, reencrypt, or passthrough routes. With cleartext, edge, or reencrypt route types, this annotation is applied as a timeout tunnel with the existing timeout value. For the passthrough route types, the annotation takes precedence over any existing timeout value set. 
+|`spec.tuningOptions.TunnelTimeout`
```
Comment (Miciah) [inline modules/nw-route-specific-annotations.adoc:80]:
```suggestion
|`spec.tuningOptions.tunnelTimeout`
```

File: modules/nw-route-specific-annotations.adoc (hunk 2 above)
Comment (Miciah) [inline modules/nw-route-specific-annotations.adoc:74]: For passthrough routes, `haproxy.router.openshift.io/timeout` applies to both client and server (see [the HAProxy configuration template](https://github.com/openshift/router/blob/b7001137f832d02d4ed68cba27fe3001ff523fa6/images/router/haproxy/conf/haproxy-config.template#L788-L789) and [the HAProxy documentation for `timeout tunnel`](http://cbonte.github.io/haproxy-dconv/2.6/configuration.html#4-timeout%20tunnel); note that passthrough routes use HAProxy's TCP mode).

File: modules/nw-route-specific-annotations.adoc [hunk 4]
```diff
@@ -16,55 +16,122 @@ To create a whitelist with multiple source IPs or subnets, use a space-delimited
...
+|
+|
+
+|`haproxy.router.openshift.io/timeout` 
+|Sets a server-side timeout for the route. (TimeUnits) 
+|`spec.tuningOptions.serverTimeout` 
+|
+
+|`haproxy.router.openshift.io/timeout-tunnel` 
+|This timeout applies to a tunnel connection, for example, WebSocket over cleartext, edge, reencrypt, or passthrough routes. With cleartext, edge, or reencrypt route types, this annotation is applied as a timeout tunnel with the existing timeout value. For the passthrough route types, the annotation takes precedence over any existing timeout value set.
```
Comment (Miciah) [inline modules/nw-route-specific-annotations.adoc:79]: I think this description could be clearer: 
* I cannot tell whether "for example" applies to "WebSocket" or to "WebSocket over cleartext, edge, reencrypt, or passthrough routes".  
* For passthrough routes, the tunnel timeout always applies (see [the HAProxy documentation for `timeout tunnel`](http://cbonte.github.io/haproxy-dconv/2.6/configuration.html#4-timeout%20tunnel)).
* I don't know what "applied as a timeout tunnel" means.  Maybe that can be deleted unless you have some idea what it means?  
* Change "types" to "type" in "the passthrough route types".  
* Instead of "takes precedence over any existing timeout value set", why not write "takes precedence over `haproxy.router.openshift.io/timeout`"?
```suggestion
|This timeout applies to a tunnel connection, such as WebSocket, over cleartext, edge, or reencrypt routes. For the passthrough route type, the annotation always applies and takes precedence over `haproxy.router.openshift.io/timeout`. 
```

File: modules/nw-route-specific-annotations.adoc [hunk 5]
```diff
@@ -16,55 +16,122 @@ To create a whitelist with multiple source IPs or subnets, use a space-delimited
...
+
+|`haproxy.router.openshift.io/timeout-tunnel` 
+|This timeout applies to a tunnel connection, for example, WebSocket over cleartext, edge, reencrypt, or passthrough routes. With cleartext, edge, or reencrypt route types, this annotation is applied as a timeout tunnel with the existing timeout value. For the passthrough route types, the annotation takes precedence over any existing timeout value set. 
+|`spec.tuningOptions.TunnelTimeout`
+|
+
+|`ingresses.config/cluster ingress.operator.openshift.io/hard-stop-after` 
+|You can set either an IngressController or the ingress config . This annotation redeploys the router and configures the HA proxy to emit the haproxy `hard-stop-after` global option, which defines the maximum time allowed to perform a clean soft-stop.
+|
+|`ingress.operator.openshift.io/hard-stop-after`
```
Comment (Miciah) [inline modules/nw-route-specific-annotations.adoc:86]: I might have misunderstoood the original bug report, but I thought one of the requests was to remove this annotation (which is *not* a *route* annotation) from the table of route annotations, and perhaps add a new procedure to describe how to use hard-stop-after.

File: modules/nw-route-specific-annotations.adoc [hunk 6]
```diff
@@ -16,55 +16,122 @@ To create a whitelist with multiple source IPs or subnets, use a space-delimited
...
+|`router.openshift.io/haproxy.health.check.interval`
+|Sets the interval for the back-end health checks. (TimeUnits) 
+|`spec.tuningOptions.healthCheckInterval` 
+|
+
 |`haproxy.router.openshift.io/ip_whitelist`
-| Sets an allowlist for the route. The allowlist is a space-separated list of IP addresses and CIDR ranges for the approved source addresses. Requests from IP addresses that are not in the allowlist are dropped.
+|Sets an allowlist for the route. The allowlist is a space-separated list of IP addresses and CIDR ranges for the approved source addresses. Requests from IP addresses that are not in the allowlist are dropped. 
+
+The maximum number of IP addresses and CIDR ranges directly visible in the `haproxy.config` file is 61. [^1^]
```
Comment (Miciah) [inline modules/nw-route-specific-annotations.adoc:96]: This is an implementation detail that doesn't make sense to mention here, especially since we're removing the environment variables (a similar implementation detail) from the documentation.

File: modules/nw-route-specific-annotations.adoc [hunk 7]
```diff
@@ -16,55 +16,122 @@ To create a whitelist with multiple source IPs or subnets, use a space-delimited
...
+|
 
-The maximum number of IP addresses and CIDR ranges directly visible in the `haproxy.config` file is 61. [^1^] |
+|`haproxy.router.openshift.io/hsts_header` 
+|Sets a Strict-Transport-Security header for the edge terminated or re-encrypt route. 
+|
+|
 
-|`haproxy.router.openshift.io/hsts_header` | Sets a Strict-Transport-Security header for the edge terminated or re-encrypt route. |
-|`haproxy.router.openshift.io/log-send-hostname` | Sets the `hostname` field in the Syslog header. Uses the hostname of the system. `log-send-hostname` is enabled by default if any Ingress API logging method, such as sidecar or Syslog facility, is enabled for the router. |
```
Comment (Miciah) [inline modules/nw-route-specific-annotations.adoc:47]: Removing this line fixes [OCPBUGS-13195](https://issues.redhat.com/browse/OCPBUGS-13195).  :tada:!  FYI @cowgirljessi.

**Review by Miciah — COMMENTED** [review]
Maybe @alebedev87 can double-check my comments on timeout settings as he has been working on an issue related to those ([OCPBUGS-14914](https://issues.redhat.com/browse/OCPBUGS-14914) / https://github.com/openshift/router/pull/536).

File: modules/nw-route-specific-annotations.adoc [hunk 8]
```diff
@@ -16,55 +16,122 @@ To create a whitelist with multiple source IPs or subnets, use a space-delimited
...
-[cols="3*", options="header"]
+[cols="4*", options="header"]
 |===
-|Variable | Description | Environment variable used as default
-|`haproxy.router.openshift.io/balance`| Sets the load-balancing algorithm. Available options are `random`, `source`, `roundrobin`, and `leastconn`.  The default value is `source` for TLS passthrough routes. For all other routes, the default is `random`. |`ROUTER_TCP_BALANCE_SCHEME` for passthrough routes. Otherwise, use `ROUTER_LOAD_BALANCE_ALGORITHM`.
-|`haproxy.router.openshift.io/disable_cookies`| Disables the use of cookies to track related connections. If set to `'true'` or `'TRUE'`, the balance algorithm is used to choose which back-end serves connections for each incoming HTTP request. |
-|`router.openshift.io/cookie_name`| Specifies an optional cookie to use for
+|Variable 
+|Description 
+|Ingress controller field
```
[the remark this answers is not in the document]
Comment (subhtk) [reply modules/nw-route-specific-annotations.adoc:23]: Let me change that. It does look clustered.

File: modules/nw-route-specific-annotations.adoc (hunk 1 above)
Comment (subhtk) [reply modules/nw-route-specific-annotations.adoc:24]: Ack.

File: modules/nw-route-specific-annotations.adoc (hunk 2 above)
Comment (subhtk) [reply modules/nw-route-specific-annotations.adoc:74]: Let me resolve this.

File: modules/nw-route-specific-annotations.adoc (hunk 7 above)
Comment (subhtk) [reply modules/nw-route-specific-annotations.adoc:47]: Great.

File: modules/nw-route-specific-annotations.adoc (hunk 4 above)
Comment (subhtk) [reply modules/nw-route-specific-annotations.adoc:79]: Done. Let me change this accordingly.

**Comment by subhtk** [comment]
Thanks for the review @Miciah. Appreciate the efforts.

File: modules/nw-route-specific-annotations.adoc (hunk 3 above)
Comment (subhtk) [reply modules/nw-route-specific-annotations.adoc:80]: ACK.

File: modules/nw-route-specific-annotations.adoc (hunk 2 above)
Comment (subhtk) [reply modules/nw-route-specific-annotations.adoc:74]: Ack.

File: modules/nw-route-specific-annotations.adoc (hunk 5 above)
Comment (subhtk) [reply modules/nw-route-specific-annotations.adoc]: @palonsoro I remember, we did have a discussion on this, whether to create a new section for hard-stop-after. Can you check again and let me know, if we have to create it.

**Comment by palonsoro** [comment]
Thanks for your comments @Miciah !!!

File: modules/nw-route-specific-annotations.adoc (hunk 5 above)
Comment (subhtk) [reply modules/nw-route-specific-annotations.adoc]: Let me remove the ingress controller field from the table and add a new section for hard-stop-after.

**Comment by ShudiLi** [comment]
/lgtm
thanks

File: modules/nw-route-specific-annotations.adoc [hunk 9]
```diff
@@ -16,95 +16,121 @@ To create a whitelist with multiple source IPs or subnets, use a space-delimited
...
+Note: Using this annotation provides basic protection against distributed denial-of-service (DDoS) attacks.
+|
+
+|`haproxy.router.openshift.io/timeout` 
+|Sets a server-side timeout for the route. (TimeUnits)
+|`spec.tuningOptions.serverTimeout` 
+
+|`haproxy.router.openshift.io/timeout-tunnel` 
+|This timeout applies to a tunnel connection, for example, WebSocket over cleartext, edge, reencrypt, or passthrough routes. With cleartext, edge, or reencrypt route types, this annotation is applied as a timeout tunnel with the existing timeout value. For the passthrough route types, the annotation takes precedence over any existing timeout value set. 
+|`spec.tuningOptions.TunnelTimeout`
```
Comment (alebedev87) [inline modules/nw-route-specific-annotations.adoc:70]:
```suggestion
|`spec.tuningOptions.tunnelTimeout`
```
To stick to `oc explain` syntax.

File: modules/nw-route-specific-annotations.adoc [hunk 10]
```diff
@@ -16,95 +16,121 @@ To create a whitelist with multiple source IPs or subnets, use a space-delimited
...
+|
+
+|`haproxy.router.openshift.io/rate-limit-connections.rate-tcp`
+|Limits the rate at which a client with the same source IP address can make TCP connections. It accepts a numeric value.  +
+Note: Using this annotation provides basic protection against distributed denial-of-service (DDoS) attacks.
+|
+
+|`haproxy.router.openshift.io/timeout` 
+|Sets a server-side timeout for the route. (TimeUnits)
+|`spec.tuningOptions.serverTimeout`
```
Comment (alebedev87) [inline modules/nw-route-specific-annotations.adoc:66]: I think that mentioning ingresscontroller fields is a good idea as it shows how to set a value for all the routes admitted by the controller. However the `tunningOptions` are not guaranteed to be specified and some default values are set by the router. These default values are often (maybe even always) mentioned in `oc explain`. Do we let the reader find them there or we want the default values to be added to the OCP docs?

File: modules/nw-route-specific-annotations.adoc (hunk 1 above)
Comment (Miciah) [reply modules/nw-route-specific-annotations.adoc]: My personal opinion is that keeping it as "Description" is better because it is shorter and because most rows don't mention an IngressController field, but maybe "Description & Ingress Controller Field" improves discoverability.

File: modules/nw-route-specific-annotations.adoc (hunk 6 above)
Comment (Miciah) [reply modules/nw-route-specific-annotations.adoc:83]: Yeah, I think the sentence about `haproxy.config` can be removed.

File: modules/nw-route-specific-annotations.adoc (hunk 10 above)
Comment (subhtk) [reply modules/nw-route-specific-annotations.adoc:66]: Shall I add note? How about this:-
"Set the spec.tuningOptions.serverTimeout field in the IngressController to apply the timeout globally.
If not set, a default value is used. To avoid confusion, always set the value explicitly.
Use oc explain to view field details, but refer to the official documentation for confirmed defaults and usage."

Comment (palonsoro) [reply modules/nw-route-specific-annotations.adoc:75]: That may be good, but also explaining what is the default value. And not only for server timeout but for the rest of the fields.

Comment (subhtk) [reply modules/nw-route-specific-annotations.adoc:75]: Yep the default will be added as well.

**Review by lpettyjo — COMMENTED** [review]
What versions does this issue apply to? The JIRA indicates 4.12-4.16. If you can indicate which versions, I'd be happy to apply the appropriate labels for the PR.

Had a couple of other comments, but otherwise, LGTM!

**Comment by palonsoro** [comment]
@lpettyjo this applies from 4.12 to 4.18. I updated the bug report.

File: modules/nw-route-specific-annotations.adoc [hunk 11]
```diff
@@ -16,95 +16,130 @@ To create an allow list with multiple source IPs or subnets, use a space-delimit
...
+`if-none`: sets the header if it is not already set.
+|`spec.httpHeaders.forwardedHeaderPolicy` 
 
 |===
 [.small]
---
-1. By default, the router reloads every 5 s which resets the balancing connection across pods from the beginning. As a result, the `roundrobin` state is not preserved across reloads. This algorithm works best when pods have nearly identical computing capabilites and storage capacity. If your application or service has continuously changing endpoints, for example, due to the use of a CI/CD pipeline, uneven balancing can result. In this case, use a different algorithm.
-2. If the number of IP addresses and CIDR ranges in an allowlist exceeds 61, they are written into a separate file that is then referenced from the `haproxy.config` file. This file is stored in the `/var/lib/haproxy/router/allowlists` folder.
+
+1. If the number of IP addresses and CIDR ranges in an allowlist exceeds 61, they are written into a separate file that is then referenced from `haproxy.config`. This file is stored in the `var/lib/haproxy/router/whitelists` folder.
```
Comment (ShudiLi) [inline modules/nw-route-specific-annotations.adoc:130]: it looks like whitelist has been changed to allowlist, so `var/lib/haproxy/router/whitelists` should be changed to `/var/lib/haproxy/router/allowlists`

File: modules/nw-route-specific-annotations.adoc [hunk 12]
```diff
@@ -16,95 +16,130 @@ To create an allow list with multiple source IPs or subnets, use a space-delimit
...
+Note: Using this annotation provides basic protection against distributed denial-of-service (DDoS) attacks.
+|
+
+|`haproxy.router.openshift.io/rate-limit-connections.rate-tcp`
+|Limits the rate at which a client with the same source IP address can make TCP connections. It accepts a numeric value.  +
+Note: Using this annotation provides basic protection against distributed denial-of-service (DDoS) attacks.
+|
+
+|`haproxy.router.openshift.io/timeout` 
+|Sets a server-side timeout for the route. If not set, the default value is `60s`. (TimeUnits)
```
Comment (ShudiLi) [inline modules/nw-route-specific-annotations.adoc:65]: The default value of  server-side timeout seems to be 30s(sorry, the https://docs.openshift.com isn't available, I will check it  again when the docs link is available)

**Review by skopacz1 — COMMENTED** [review]
I think before I can even fully review this PR, the one comment I made about the column number should be addressed so I can better view the table and understand what's being changed here

**Comment by skopacz1** [comment]
Let's maybe work through this PR together on Slack, in the meantime I am going to remove merge review labels because I am no longer on squad

**Comment by lihongan** [comment]
/retitle OCPBUGS-19700: Modified Route-specific annotations table and added a section for ingress controller annotation

**Comment by kalexand-rh** [comment]
This PR was replaced by https://github.com/openshift/openshift-docs/pull/98192.
